### DISCUSSION TITLE
Disable bluetooth at least temporarily (e.g. for competition purposes)

### DISCUSSION THREAD

**Discussion by DrTom** [pybricks/support#1089] [Q&A] [answered]
We are planning to use Pybricks for the WRO RoboMission. 

Having bluetooth enabled permanently is problematic for competition purposes. The rules (Section 5.11 in https://wro-association.org/wp-content/uploads/WRO-2023-RoboMission-General-Rules.pdf) cause some inconveniences and there is the general problem that a connection is easy to hijack which can be used to ruin a run of a competitor. While that is forbidden it would be very difficult to pinpoint an attacker.

The cleanest way to resolve this would be to disable bluetooth (at least temporarily). It seems to me that this is one of the few things which are actually better with the lego firmeware which only enables bluetooth after the dedicated hardware button is pressed. 

1. Is there a way to disable bluetooth with the current firmeware? 
2. If not so: is there an easy way patch the pypricks code and build a custom firmeware which achieves that? 
3. Are there any plans for the future to implement this?

**Comment by dlech** [comment]
> 1\. Is there a way to disable bluetooth with the current firmeware?

No.

> 2\. If not so: is there an easy way patch the pypricks code and build a custom firmeware which achieves that?

Change `PBSYS_CONFIG_BLUETOOTH` to `0` and build the firmware. But with no Bluetooth, there is no way to transfer a program (no USB support). On hubs with external flash, the previously program saved on power off will still be available though.

> 3\. Are there any plans for the future to implement this?

No plans, but feel free to open a feature request detailing the use case and proposed solution.

**Comment by laurensvalk** [comment]
This occasionally comes up. Please do tell your WRO organizer about Pybricks and its possibilities.

If there are enough teams that want this feature, this is something we could potentially support.

**Comment by DrTom** [comment]
First thank you for this quick clarification.  

> Change PBSYS_CONFIG_BLUETOOTH to 0 and build the firmware. But with no Bluetooth, there is no way to transfer a program (no USB support). On hubs with external flash, the previously program saved on power off will still be available though.

Right, so simply disabling bluetooth will be of no use. I believe most if not all bluetooth enabled Lego devices do not have an external flash. 

Next best suggestion I see: keep bluetooth on by default but use the bluetooth button to toggle it. In this way the program could be transferred and then bluetooth disabled for the actual run. Not perfect but much better than the current state. 

Could you give some hints how difficult it is to achieve this? I am halfway comfortable with Micropython (so far only on ESP32 devices); `C` would probably be too hard. I have not seen any bluetooth documentation for the Pybricks Python API;  are there possibly undocumented features? 

I will also try to raise attention at the WRO (though as a coach my influence is limited).

**Reply by dlech** [reply]
> I believe most if not all bluetooth enabled Lego devices do not have an external flash.

the SPIKE and Robot Inventor hubs do

**Reply by dlech** [reply]
> Could you give some hints how difficult it is to achieve this?

This requires C programming to modify the firmware. As a quick hack, you could add a "wait for button press" [here](https://github.com/pybricks/pybricks-micropython/blob/d80df1a4c9b23853e274753fd009d25bf412ff75/lib/pbio/sys/bluetooth.c#L283-L288) to avoid powering on the bluetooth chip and advertising until the Bluetooth button is pressed. It would probably be a good enough solution if your team isn't using Bluetooth in your programs but more thought would be needed for a general solution that works for all users.

**Comment by DrTom** [comment]
> the SPIKE and Robot Inventor hubs do

Hm, possibly there is some misunderstanding of an "external flash" on my side. I thought of something like an SD Card. When I proposed disabling bluetooth I though of using USB to transfer the program. I was ignorant this is not even possibly (we never tried so far). Anyways reflashing the firmeware during the competition is pretty much out of the question.

**Comment by DrTom** [comment]
>  As a quick hack, you could add a "wait for button press" 

That would be a good workaround for the time being. I will have a look into it.

**Comment by Kermit647** [comment]
Is there a tutorial how to compile/upload a custom firmware somewhere ?

**Reply by dlech** [reply]
https://pybricks.com/install/technic-boost-city/#installing-the-latest-build-advanced

**Comment by Kermit647** [comment]
I made a custom firmware for this purpose. I hacked the bluetooth like dlech suggested. It will now only enable bluetooth after you press the bluetooth button once. Works like expected.

**Reply by dlech** [reply]
Great! 

If you make a draft pull request, our CI will build it and upload firmware.zip files that others can try too without having to build it themselves.

**Reply by Kermit647** [reply]
I just dirty hacked in a global variable, thats not realy worthy to publish

**Comment by DrTom** [comment]
@Kermit647 Is your code available on GitHub e.g.?

**Reply by Kermit647** [reply]
I mape a upload from the firmware for you to try out :
[http://stauff.org/firmware.zip](http://stauff.org/firmware.zip)

**Reply by ttn-c4plus** [reply]
Hello, I used it, and it work. So can you change PBSYS_CONFIG_BLUETOOTH to 0 and build the firmware again for me. I use it for WRO. Thank you

**Reply by Kermit647** [reply]
If you you change PBSYS_CONFIG_BLUETOOTH to 0 you will have no bluetooth at all and will not be able to upload anything

**Reply by ttn-c4plus** [reply]
@Kermit647 are you free? Can I get your contact? I have a question about build firmware to use Bluetooth.
If you use facebook: [Me](https://www.facebook.com/ttn.c4plus)
Hope you help!

**Comment by afarago** [comment]
Hi!

I am contributing to WRO on the organizer side and in the countries I am judging, or acting a s a head judge I always emphasize the importance of focusing on giving the maximum fun and learning experience to the kids.

Several WRO rules are impossible to adhere to - here comes the creative flexibility and joint teamwork of the kids & coaches & judges.

- SPIKE V2 Bluetooth cannot be turned off (although you use USB - BT is always on)
- SPIKE V3 - same BT on issue, also there is no way to delete programs from slots, thus the rule to have only one program on the bot needs to be creatively taken. (We ask which slot the kids will use and take a not on it)
- SPIKE V2 starting with center button introduces 1.8sec unfair delay due to the animation & sounds.

Hope this helps you and the kids. WRO judges are typically nice person's ;) Just make sure you talk with them.
Feel free to reach out in case you need help.

**Comment by DrTom** [comment]
Thank you all for your input. Time was running out for me to prepare a custom firmware for this years competition. As @afarago mentioned there exceptions in the regulations.
I still believe there should be a sustainable solution for this problem in the future. So I am leaving this ticket open.

**Reply by afarago** [reply]
Yes, absolutely awesome.
Let's hope we can connect pybricks with the major competition heads such as WRO and FLL.

In the meantime, just reach out to the head judge of the event and talk with them. 99% you will find an agreement.

**Comment by nkarstens** [comment]
I'm going to take a shot at implementing this. Please see https://github.com/pybricks/pybricks-micropython/pull/195.

**Comment by ggramlich** [comment]
Hello everybody,

we had our regional WRO Elementary competition yesterday in Germany. It was the first weekend of Elementary regionals, there are still a lot of them coming up. Out of the 35 Elementary teams which competed yesterday, the top 4 teams  in Germany (two teams from another coach and my two teams) were in our regional competition in Bad Homburg and they were all using Pybricks. Of course correlation is not causation and it might just be that the more ambitious teams are looking for alternatives to the Lego software more often than the others.
The other coach made me aware of this thread and the availabality of the patched firmware to turn off bluetooth, which they were using.
As we informed the judges beforehand that we could not turn bluetooth off, it was actually no problem in the regionals and it should not be a problem in the German Finals. But being able to turn it off could prevent discussions and misunderstandings.

@dlech @laurensvalk 
It would be awesome to have https://github.com/pybricks/pybricks-micropython/pull/195 (and possibly also https://github.com/pybricks/pybricks-micropython/pull/208) integrated into the firmware. Special thanks to @nkarstens for providing the pull request.

We are more than willing to test it in the beta version 😊

**Reply by laurensvalk** [reply]
Thank you @ggramlich for this feedback.

I was originally thinking to combine the Bluetooth disabling with USB support, but I think we can expedite the Bluetooth toggle separately. I'll open a dedicated issue for it. Thanks!

**Reply by nkarstens** [reply]
@ggramlich Glad you found it to be useful!

**Comment by laurensvalk** [comment, marked as answer]
This feature is now tracked as an actionable task in https://github.com/pybricks/support/issues/1615.

Thanks for the feedback! We'll close this discussion here. Please feel free to add suggestions to the link above.
